source/common/network/apple_dns_impl.h

```diff
@@ -0,0 +1,95 @@
+public:
+  AppleDnsResolverImpl(Event::Dispatcher& dispatcher);
+  ~AppleDnsResolverImpl() override;
+
+  // Network::DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override;
+
+private:
+  struct PendingResolution : public ActiveDnsQuery {
+    PendingResolution(AppleDnsResolverImpl& parent, ResolveCb callback,
+                      Event::Dispatcher& dispatcher, DNSServiceRef sd_ref,
+                      const std::string& dns_name)
+        : parent_(parent), callback_(callback), dispatcher_(dispatcher), individual_sd_ref_(sd_ref),
+          dns_name_(dns_name) {}
+    ~PendingResolution();
+
+    // Network::ActiveDnsQuery
+    void cancel() override;
+
+    static DnsResponse buildDnsResponse(const struct sockaddr* address, uint32_t ttl);
+    // Wrapper for the API call.
+    DNSServiceErrorType dnsServiceGetAddrInfo(DnsLookupFamily dns_lookup_family);
+    // Wrapper for the API callback.
+    void onDNSServiceGetAddrInfoReply(DNSServiceFlags flags, uint32_t interface_index,
+                                      DNSServiceErrorType error_code, const char* hostname,
+                                      const struct sockaddr* address, uint32_t ttl);
+
+    // Small wrapping struct to accumulate addresses from firings of the
+    // onDNSServiceGetAddrInfoReply callback.
+    struct FinalResponse {
+      ResolutionStatus status_;
+      std::list<DnsResponse> responses_;
+    };
+
+    AppleDnsResolverImpl& parent_;
+    // Caller supplied callback to invoke on query completion or error.
+    const ResolveCb callback_;
+    // Dispatcher to post any callback_ exceptions to.
+    Event::Dispatcher& dispatcher_;
+    DNSServiceRef individual_sd_ref_;
+    const std::string dns_name_;
+    bool synchronously_completed_{};
+    bool owned_{};
+    // DNSServiceGetAddrInfo fires one callback DNSServiceGetAddrInfoReply callback per IP address,
+    // and informs via flags if more IP addresses are incoming. Therefore, these addresses need to
+    // be accumulated before firing callback_.
+    absl::optional<FinalResponse> pending_cb_{};
+  };
+
+  void initializeMainSdRef();
+  void deallocateMainSdRef();
+  void onEventCallback(uint32_t events);
+  void addPendingQuery(PendingResolution* query) { queries_with_pending_cb_.push_back(query); }
+  void flushPendingQueries();
+
+  Event::Dispatcher& dispatcher_;
+  DNSServiceRef main_sd_ref_;
+  Event::FileEventPtr sd_ref_event_;
+  // When using a shared sd ref via DNSServiceCreateConnection, the DNSServiceGetAddrInfoReply
+  // callback with the kDNSServiceFlagsMoreComing flag might refer to addresses for various
+  // PendingResolutions. Therefore, the resolver needs to have a container of queries pending
+  // calling their own callback_s until a DNSServiceGetAddrInfoReply is called with
+  // kDNSServiceFlagsMoreComing not set.
+  std::list<PendingResolution*> queries_with_pending_cb_;
```

Comment:
Maybe use a set, to avoid a linear scan. Also, do we dedup pending queries?

Reply:
Yep, they are dedup here

 https://github.com/envoyproxy/envoy/pull/13074/files#diff-4ebdef3732b2b07e175aba41bcaf62abR193

and here for the error case: 

https://github.com/envoyproxy/envoy/pull/13074/files#diff-4ebdef3732b2b07e175aba41bcaf62abR156

Reply:
re: list. I actually updated the logic on cancellation, which was previously wrong, which causes linear scans to be a moot point. However, I did move to a set, because I want to assert absence in the container when adding so I still benefit from the set in non-release builds.